File: apps/api/app/schemas/agents.py
from enum import Enum
from typing import Literal

from pydantic import BaseModel, Field


class TrackType(str, Enum):
    STRETCH = "Stretch"
    REALISTIC = "Realistic"
    SAFE = "Safe"


class AiRiskTier(int, Enum):
    TAILWIND = 1
    AUGMENTED = 2
    PARTIAL = 3
    HEAVY = 4
    SUNSET = 5


class CareerTrack(BaseModel):
    name: str
    career_slug: str
    type: TrackType
    why_recommended: str
    avg_starting_salary_band: str
    time_to_job_estimate_months: int = Field(ge=1, le=36)
    competition_level: Literal["low", "medium", "high"]
    top_hiring_companies: list[str] = Field(default_factory=list, max_length=8)
    honest_warning: str | None = None
    ai_risk_tier: AiRiskTier | None = None
    ai_risk_label: str | None = None
    replaces: list[str] = Field(default_factory=list)
    amplifies: list[str] = Field(default_factory=list)
    evolved_role_2029: str | None = None
    survival_skills: list[str] = Field(default_factory=list)


class CareerPathPickerOutput(BaseModel):
    tracks: list[CareerTrack] = Field(min_length=3, max_length=3)


class DisruptionBriefing(BaseModel):
    summary: str
    rejected_paths: list[str] = Field(default_factory=list)
    do_not_pursue_callouts: list[str] = Field(default_factory=list)


class AiRealityCheckOutput(BaseModel):
    tracks: list[CareerTrack] = Field(min_length=3, max_length=3)
    disruption_briefing: DisruptionBriefing


class SkillGapItem(BaseModel):
    name: str
    current_level: int = Field(ge=0, le=5)
    required_level: int = Field(ge=1, le=5)
    demand_frequency_pct: float = Field(ge=0, le=100)
    weeks_to_bridge: int = Field(ge=0, le=52)
    priority: Literal["must", "should", "nice"]


class SkillGapAnalyzerOutput(BaseModel):
    role: str
    career_slug: str
    skills: list[SkillGapItem]


class LearningWeek(BaseModel):
    week: int = Field(ge=1)
    focus_skill: str
    resources: list[str]
    hours: int = Field(ge=1)
    mini_task: str
    checkpoint: bool = False


class LearningPathGeneratorOutput(BaseModel):
    total_weeks: int
    weeks: list[LearningWeek]


class PortfolioProject(BaseModel):
    name: str
    difficulty: Literal["Beginner", "Easy", "Medium", "Hard", "Capstone"]
    problem_statement: str
    dataset_or_api: str | None = None
    expected_output: str
    hours_estimate: int = Field(ge=4)
    tech_stack: list[str]
    resume_bullet: str


class ProjectIdeatorOutput(BaseModel):
    projects: list[PortfolioProject] = Field(min_length=5, max_length=5)


class CertificationItem(BaseModel):
    name: str
    provider: str
    cost: str
    hours: int
    why: str


class CertificationAdvisorOutput(BaseModel):
    recommended: list[CertificationItem]
    skip: list[dict[str, str]]  # {name, why_skip}


class PortfolioSiteSection(BaseModel):
    section: Literal["hero", "about", "projects", "skills", "contact"]
    content: str


class LinkedInPost(BaseModel):
    week: int
    post_type: Literal["project", "learning", "opinion", "resource"]
    template: str


class PortfolioBuilderOutput(BaseModel):
    github_readme_md: str
    portfolio_site_sections: list[PortfolioSiteSection]
    hosting_suggestion: str
    linkedin_calendar: list[LinkedInPost]
    connection_request_templates: list[str] = Field(min_length=5, max_length=5)


class AgentOutputs(BaseModel):
    career_paths: CareerPathPickerOutput | None = None
    ai_reality_check: AiRealityCheckOutput | None = None
    skill_gap: SkillGapAnalyzerOutput | None = None
    learning_path: LearningPathGeneratorOutput | None = None
    projects: ProjectIdeatorOutput | None = None
    certifications: CertificationAdvisorOutput | None = None
    portfolio: PortfolioBuilderOutput | None = None
    chosen_track_type: TrackType | None = None

File: apps/web/src/lib/schemas/agents.ts
import { z } from "zod";

export const trackTypeSchema = z.enum(["Stretch", "Realistic", "Safe"]);
export const aiRiskTierSchema = z.union([
  z.literal(1),
  z.literal(2),
  z.literal(3),
  z.literal(4),
  z.literal(5),
]);

export const careerTrackSchema = z.object({
  name: z.string(),
  career_slug: z.string(),
  type: trackTypeSchema,
  why_recommended: z.string(),
  avg_starting_salary_band: z.string(),
  time_to_job_estimate_months: z.number().int().min(1).max(36),
  competition_level: z.enum(["low", "medium", "high"]),
  top_hiring_companies: z.array(z.string()).max(8).default([]),
  honest_warning: z.string().nullable().optional(),
  ai_risk_tier: aiRiskTierSchema.nullable().optional(),
  ai_risk_label: z.string().nullable().optional(),
  replaces: z.array(z.string()).default([]),
  amplifies: z.array(z.string()).default([]),
  evolved_role_2029: z.string().nullable().optional(),
  survival_skills: z.array(z.string()).default([]),
});

export const careerPathPickerOutputSchema = z.object({
  tracks: z.tuple([
    careerTrackSchema,
    careerTrackSchema,
    careerTrackSchema,
  ]),
});

export const disruptionBriefingSchema = z.object({
  summary: z.string(),
  rejected_paths: z.array(z.string()).default([]),
  do_not_pursue_callouts: z.array(z.string()).default([]),
});

export const aiRealityCheckOutputSchema = z.object({
  tracks: z.tuple([
    careerTrackSchema,
    careerTrackSchema,
    careerTrackSchema,
  ]),
  disruption_briefing: disruptionBriefingSchema,
});

export const skillGapItemSchema = z.object({
  name: z.string(),
  current_level: z.number().int().min(0).max(5),
  required_level: z.number().int().min(1).max(5),
  demand_frequency_pct: z.number().min(0).max(100),
  weeks_to_bridge: z.number().int().min(0).max(52),
  priority: z.enum(["must", "should", "nice"]),
});

export const skillGapAnalyzerOutputSchema = z.object({
  role: z.string(),
  career_slug: z.string(),
  skills: z.array(skillGapItemSchema),
});

export const learningWeekSchema = z.object({
  week: z.number().int().min(1),
  focus_skill: z.string(),
  resources: z.array(z.string()),
  hours: z.number().int().min(1),
  mini_task: z.string(),
  checkpoint: z.boolean().default(false),
});

export const learningPathGeneratorOutputSchema = z.object({
  total_weeks: z.number().int().min(1),
  weeks: z.array(learningWeekSchema),
});

export const portfolioProjectSchema = z.object({
  name: z.string(),
  difficulty: z.enum([
    "Beginner",
    "Easy",
    "Medium",
    "Hard",
    "Capstone",
  ]),
  problem_statement: z.string(),
  dataset_or_api: z.string().nullable().optional(),
  expected_output: z.string(),
  hours_estimate: z.number().int().min(4),
  tech_stack: z.array(z.string()),
  resume_bullet: z.string(),
});

export const projectIdeatorOutputSchema = z.object({
  projects: z
    .tuple([
      portfolioProjectSchema,
      portfolioProjectSchema,
      portfolioProjectSchema,
      portfolioProjectSchema,
      portfolioProjectSchema,
    ])
    .or(z.array(portfolioProjectSchema).length(5)),
});

export const certificationItemSchema = z.object({
  name: z.string(),
  provider: z.string(),
  cost: z.string(),
  hours: z.number().int(),
  why: z.string(),
});

export const certificationAdvisorOutputSchema = z.object({
  recommended: z.array(certificationItemSchema),
  skip: z.array(
    z.object({
      name: z.string(),
      why_skip: z.string(),
    }),
  ),
});

export const portfolioBuilderOutputSchema = z.object({
  github_readme_md: z.string(),
  portfolio_site_sections: z.array(
    z.object({
      section: z.enum(["hero", "about", "projects", "skills", "contact"]),
      content: z.string(),
    }),
  ),
  hosting_suggestion: z.string(),
  linkedin_calendar: z.array(
    z.object({
      week: z.number().int(),
      post_type: z.enum(["project", "learning", "opinion", "resource"]),
      template: z.string(),
    }),
  ),
  connection_request_templates: z.array(z.string()).min(5).max(5),
});

export const agentOutputsSchema = z.object({
  career_paths: careerPathPickerOutputSchema.nullable().optional(),
  ai_reality_check: aiRealityCheckOutputSchema.nullable().optional(),
  skill_gap: skillGapAnalyzerOutputSchema.nullable().optional(),
  learning_path: learningPathGeneratorOutputSchema.nullable().optional(),
  projects: projectIdeatorOutputSchema.nullable().optional(),
  certifications: certificationAdvisorOutputSchema.nullable().optional(),
  portfolio: portfolioBuilderOutputSchema.nullable().optional(),
  chosen_track_type: trackTypeSchema.nullable().optional(),
});

export type TrackType = z.infer<typeof trackTypeSchema>;
export type CareerTrack = z.infer<typeof careerTrackSchema>;
export type AiRealityCheckOutput = z.infer<typeof aiRealityCheckOutputSchema>;
export type AgentOutputs = z.infer<typeof agentOutputsSchema>;
export type IntakeFormAgents = AgentOutputs;
